x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverter.java

```diff
@@ -148,10 +164,71 @@ public static TemporalAmount parseTemporalAmout(Number value, String qualifier,
         };
     }
 
+    /**
+     * Converts a string or date expression to long.
+     */
+    private static Long foldExpressionToLong(Expression exp) {
+        Object value = Foldables.valueOf(exp);
+        if (DataTypes.isDateTime(exp.dataType())) {
+            return ((Number) value).longValue();
+        } else if (DataTypes.isString(exp.dataType())) {
+            return convertDatetimeStringToLong(((BytesRef) value).utf8ToString(), ESQL_DEFAULT_DATE_TIME_FORMATTER);
+        }
+        throw new IllegalArgumentException("unsupported type [" + exp.dataType() + "]");
```

Comment:
shouldn't this be null like the method below (to ChronoField)?
Otherwise the two are inconsistent.